Unify error messages and add Troubleshooting section to docs (#4915)
software-mansion/react-native-reanimated: docs/docs/guides/troubleshooting.mdx

---
id: Troubleshooting
title: Troubleshooting
sidebar_label: Troubleshooting

> sidebar_position: 1

Reviewer: ```suggestion
sidebar_position: 999
```

---

# Troubleshooting

> ## Version mismatching

Reviewer: ```suggestion
## Initialization issues
```

Reanimated has four core components that compose its code:

- C++,
- Java,
- JavaScript,
- Reanimated Babel plugin.

All of them are supposed to work correctly only within the same minor version. Therefore, having any of those pieces in your code - directly or indirectly - separate from `react-native-reanimated`, especially having any code transpiled with a different version of the aforementioned plugin will result in undefined behavior and errors.

### Unknown version of Reanimated Babel plugin

> This can happen when:

Reviewer: ```
**Problem:** Lorem ipsum sit dolor amet.

**Solution:** Lorem ipsum sit dolor amet.
```

---
id: Troubleshooting
title: Troubleshooting
sidebar_label: Troubleshooting
---

# Troubleshooting

## Initialization issues

Reanimated has four core components that compose its code:

- C++,
- Java,
- JavaScript,
- Reanimated Babel plugin.

All of them are supposed to work correctly only within the same minor version. Therefore, having any of those pieces in your code - directly or indirectly - separate from `react-native-reanimated`, especially having any code transpiled with a different version of the aforementioned plugin will result in undefined behavior and errors.

### Failed to create a worklet

**Problem:** This usually happens when Reanimated is not properly installed, e.g. forgetting to include the Reanimated Babel plugin in `babel.config.js`.

**Solution:** See installation docs at https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/installation#babel-plugin for more information.

### Native part of Reanimated doesn't seem to be initialized

> **Problem:**
> This issue happens when Reanimated fails to initialize its native side from JavaScript.

Reviewer: ```suggestion
**Problem:** This issue happens when Reanimated fails to initialize its native side from JavaScript.
```

**Solution:**

1. If you recently installed or upgraded Reanimated, make sure to rebuild your app.
2. Check if your platform is supported by Reanimated. Currently we support:
   - Android

>    - iOS

Reviewer: ```suggestion
   - iOS
   - macOS
```

   - Web

> 3. If you are running in a brownfield app, make sure to initialize the native library manually.

Reviewer: ```suggestion
3. If you are using Reanimated in a brownfield app, make sure to initialize the native library manually.
```

**Solution:** Modify your `package.json` file accordingly:

- if you use `yarn`, you should add [`resolution` property](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/):

```json
"resolutions": {
  "react-native-reanimated": <Reanimated version>
}
```

- if you use `npm`, you should add [`overrides` property](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#overrides):

```json
"overrides": {
  "react-native-reanimated": <Reanimated version>
}
```

After that make sure to run your package manager again, either `yarn` or `npm install`.

### Outdated version of `react-native` for New Architecture

**Problem:** Reanimated supports New Architecture (Fabric) only on the latest minor release of React Native.

> **Solution:** Please upgrade to at React Native 0.72 or downgrade to an older version of Reanimated 3.

Reviewer: ```suggestion
**Solution:** Please upgrade to at least React Native 0.72 or downgrade to an older version of Reanimated 3.
```

## Threading issues

### Tried to synchronously call a non-worklet function on the UI thread

**Problem:** This can happen when you try to call a function that is not marked as a worklet from a worklet. E.g.:

```typescript
function callee() {
  console.log('hello');
}
function caller() {
  'worklet';
  callee(); // <- this will throw in `runOnUI`
}
runOnUI(caller)();
```

In this example, `callee` cannot be called from a worklet ran on UI thread because there is no corresponding UI function for `callee`.

**Solution:**

1. If you want to synchronously execute this method, mark it as a worklet using `worklet` directive:

```diff

> function callee(){
> +   'worklet';
>     console.log("hello");
>   }

Reviewer: ```suggestion
 function callee() {
+  'worklet';
   console.log("hello");
 }
```

```

2. If you want to execute this function on the JS thread, wrap it using `runOnJS`:

```diff

> function caller(){
>   'worklet';
> -   callee();
> +   runOnJS(callee)();
> }

Reviewer: ```suggestion
 function caller() {
   'worklet';
-  callee();
+  runOnJS(callee)();
 }
```
